src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/DomainEventServiceTest.kt

```diff
@@ -0,0 +1,126 @@
+  fun `getApplicationSubmittedDomainEvent returns null when no event does not exist`() {
+    val id = UUID.fromString("c3b98c67-065a-408d-abea-a252f1d70981")
+
+    every { domainEventRespositoryMock.findByIdOrNull(id) } returns null
+
+    assertThat(domainEventService.getApplicationSubmittedDomainEvent(id)).isNull()
+  }
+
+  @Test
+  fun `getApplicationSubmittedDomainEvent returns event`() {
+    val id = UUID.fromString("c3b98c67-065a-408d-abea-a252f1d70981")
+    val applicationId = UUID.fromString("a831ead2-31ae-4907-8e1c-cad74cb9667b")
+    val occurredAt = OffsetDateTime.parse("2023-02-01T14:03:00+00:00")
+    val crn = "CRN"
+
+    val data = ApplicationSubmittedEnvelope(
+      id = id,
+      timestamp = occurredAt,
+      eventType = "approved-premises.application.submitted",
+      eventDetails = ApplicationSubmittedFactory().produce()
+    )
+
+    every { domainEventRespositoryMock.findByIdOrNull(id) } returns DomainEventEntityFactory()
+      .withId(id)
+      .withApplicationId(applicationId)
+      .withCrn(crn)
+      .withType(DomainEventType.APPROVED_PREMISES_APPLICATION_SUBMITTED)
+      .withData(objectMapper.writeValueAsString(data))
+      .withOccurredAt(occurredAt)
+      .produce()
+
+    val event = domainEventService.getApplicationSubmittedDomainEvent(id)
+    assertThat(event).isEqualTo(
+      DomainEvent(
+        id = id,
+        applicationId = applicationId,
+        crn = "CRN",
+        occurredAt = occurredAt,
+        data = data
+      )
+    )
+  }
+
+  @Test
+  fun `save throws for unrecognised event type`() {
+    val exception = assertThrows<RuntimeException> {
+      domainEventService.save(
+        DomainEvent<String>(
+          id = UUID.fromString("1a520974-d5d6-49a4-9ce1-827427dd489f"),
+          applicationId = UUID.fromString("ea100c7f-2e16-4280-b1b2-d2f666a6dcd1"),
+          crn = "CRN123",
+          occurredAt = OffsetDateTime.now(),
+          data = "some data"
+        )
+      )
+    }
+
+    assertThat(exception.message).isEqualTo("Unrecognised domain event type: java.lang.String")
+  }
+
+  @Test
+  fun `save persists event`() {
+    val id = UUID.fromString("c3b98c67-065a-408d-abea-a252f1d70981")
+    val applicationId = UUID.fromString("a831ead2-31ae-4907-8e1c-cad74cb9667b")
+    val occurredAt = OffsetDateTime.parse("2023-02-01T14:03:00+00:00")
+    val crn = "CRN"
+
+    every { domainEventRespositoryMock.save(any()) } answers { it.invocation.args[0] as DomainEventEntity }
+
+    domainEventService.save(
+      DomainEvent(
+        id = id,
+        applicationId = applicationId,
+        crn = crn,
+        occurredAt = OffsetDateTime.now(),
+        data = ApplicationSubmittedEnvelope(
+          id = id,
+          timestamp = occurredAt,
+          eventType = "approved-premises.application.submitted",
+          eventDetails = ApplicationSubmittedFactory().produce()
+        )
+      )
+    )
+
+    verify(exactly = 1) { domainEventRespositoryMock.save(any()) }
```

Comment:
Should we flesh out this "verification" to demonstrate for example that the repository receives a `DomainEventEntity` where:

-  the `data` attribute is a JSON version of `application.submitted` and 
- that `type` is the expected `DomainEventType.APPROVED_PREMISES_APPLICATION_SUBMITTED`

Or will that come later / be covered elsewhere?

Reply:
Good point!  Have updated the verification.